### PR TITLE
ENH: allow int-like parameters in several routines

### PR DESCRIPTION
**Pull request by pv**
This is a follow-up to gh-7351 taking the operator.index approach.

Afaik this is backward compatible, apart from the checks added in resample_poly. operator.index accepts anything that can be safely cast to a Python integer scalar. This includes stuff like numpy integers, 0d numpy arrays etc. --- but not floating point inputs which is intentional.

- Replace `isinstance(n, int)` checks by `n = operator.index`, making them more lenient.
- Factor out `floor(n) != n` to a helper function in scipy.special.
- Add integer-valuedness check to resample_poly (more of a bugfix, doesn't require deprecation cycle imho)

### CODE REVIEW

**Comment by ev-br** [comment]
`operator.index` approach looks great, should probably be extended to other int-ness checks across the codebase, here or in a separate PR. I'll wait a bit to see if there are comments from folks more familiar with `fftpack` and `signal`.

**Review by larsoner — APPROVED** [review]
I'm happy with the `fftpack` and `signal` changes

**Comment by endolith** [comment]
Oh yeah, I was going to bring this up on the mailing list and argue that "duck-valuing" actually *is* the correct way to do things, but I never got around to it. :D  

Basically that these are not indexes to arrays, and so `operator.index` is inappropriate, and see also PEP 238, etc:

> In a unified numeric model, the different types (int, long, float, complex, and possibly others, such as a new rational type) serve mostly as storage optimizations, and to some extent to indicate orthogonal properties such as inexactness or complexity. In a unified model, **the integer 1 should be indistinguishable from the floating point number 1.0** (except for its inexactness), and both should behave the same in all numeric contexts.

(But if all this PR does is *increase* the amount of objects that these functions accept, then I'm fine with it.  We can argue about whether it should accept even more things at another time.)

**Comment by pv** [comment]
Yes, I think this PR is currently partly independent of this discussion
as it just relaxes type checks.

However, note that the PEP 238 statement says "indistinguishable ...
(except for its inexactness)". I would argue that the uses of integers
here are not in the general numeric context, as integer exactness of the
values is *required*.

Moreover, note that in PEP357 the motivation for `__index__` is not only
about indices --- indeed, PEP357 says "__index__ special method so that
arbitrary objects can be used whenever integers are explicitly needed in
Python".

**Comment by endolith** [comment]
Ok, I found my draft email.  :)

Everyone agrees that the function should be modified to accept numpy ints.  The question is whether to also accept things like integer-valued floats.  Examples:

    int(5)      passes int(n) == n, has n.__index__() method, passes isinstance(n, numbers.Integral)
    np.int64(5) passes int(n) == n, has n.__index__() method, passes isinstance(n, numbers.Integral)
    array(5)    passes int(n) == n, has n.__index__() method
    float(5)    passes int(n) == n
    Fraction(5) passes int(n) == n
    Decimal(5)  passes int(n) == n

So for example, I think this realistic example should work:

    audio_fs = 48000
    control_fs = 1000
    q = audio_fs / control_fs
    decimate(data, q)

Here the sampling rates are int type, but Python 3 coerces q to a float 48.0, but it still has the exact integer value intended and makes sense to be accepted in this context.  Then if `audio_fs` changes to 44100, decimate will raise an exception, something like:

    "ValueError: q is 44.1.  Must be a non-negative integer".

pointing out exactly what the problem is and forcing the user to decide how they want to handle this. (Most likely, they need to change to a completely different function like `resample_poly()`.)

`operator.index` was introduced to avoid silent truncation errors in slicing, but using it in this context would *introduce* silent truncation errors:  If instead, `decimate(data, q)` produced an exception like

    "TypeError: q must be of integer type",

we force users to do something like

    q = int(audio_fs / control_fs)

and then if `audio_fs` is later changed to 44100, they'll be bitten by silent truncation that they hadn't thought of. 

In other words, especially when using `decimate()`, it's likely the case that *none* of `int()`, `ceil()`, or `floor()` are appropriate, but that's what people will first reach for when they get a TypeError.  So "forcing the user to deal with it early on" isn't a good idea.  The right thing to do is raise an Exception if the argument to `decimate` does not have integer value.

(This is similar to the rationale for why `/` was changed to return a float even if the inputs are ints, and why `math.ceil()` was changed to return an int even if the input is float.  In Python, it's the value that matters, not the type.)

**Comment by larsoner** [comment]
@endolith can you open a separate issue for further enhancements? If we go your suggested route it can be done in a subsequent PR, right?

**Comment by perimosocordiae** [comment]
I'm +1 to merge this as-is.

**Review by endolith — APPROVED** [review]
Yes, this looks good to me

**Comment by endolith** [comment]
> @endolith can you open a separate issue for further enhancements? If we go your suggested route it can be done in a subsequent PR, right?

I'll just leave it as https://github.com/scipy/scipy/pull/7351 and rebase it on top of this?

**Comment by larsoner** [comment]
I forgot about #7351. That makes sense to me.

@pv do you see #7351 as a reasonable follow-up to this PR?

**Comment by endolith** [comment]
Well this is a follow-up to that, so pv would need to remove the "closes" comment so it doesn't automatically close? https://github.com/scipy/scipy/pull/7690#issue-246609198  (or I can re-open?)

**Comment by rgommers** [comment]
All in agreement, so merging. Thanks @pv, all.

**Comment by rgommers** [comment]
gh-7351 is still open, @endolith go ahead and rebase that I'd say.

**Comment by endolith** [comment]
@pv in your checks for int-ness, why did you compare to `floor(n)` instead of `int(n)`?

**Comment by pv** [comment]
Yes, int would have been better since it doesn't lose precision (due to the limited number of bits in fp mantissa) if the argument is already int.
